### PR TITLE
[Flax] Add FlaxBart models

### LINKED ISSUE #11478: [Flax] Add FlaxBart model
# 🚀 Feature request

It would be nice to implement Flax version of BART model.

## Motivation

Narrow the gap in support between encoder transformers (BERT, RoBERTa, ...) and encoder-decoder models (BART,...).

## Your contribution

I've been working on this now so I hope I will send a PR soon.

@patrickvonplaten @sgugger

### PR DESCRIPTION
**Pull request by stancld**
# What does this PR do?

This PR adds Flax implementation of BART and classes for various downstream tasks. Fixes #11478.

Most of the code is inspired by the Flax implementation of BERT and PyTorch implementation of BART.

From Suraj:
A couple of important points to note:

- The seq2seq API is slightly different from the PyTorch BART model in that the `__call__` method of `FlaxBART` does not accept `encoder_outputs`. In PT if `encoder_outputs` is passed then the encoder is skipped and only the decoder is called. This is not supported in `FlaxBART` to prevent any unintended issues during JIT compilation as skipping module or passing different inputs to function causes re-compilation. Also, the idiomatic way of accessing intermediate modules in Flax models is to expose explicit methods.  So the API is as follows 
   - the `__call__` method expects both the encoder and decoder inputs and does forward pass through both the modules
   - Every model has a `encode` and `decode` method which should be called if one wants to run just the encoder or decoder. The `decode` method only returns the decoder outputs and for `*ForConditionalGeneration` modules it also returns the `logits`
   
```python
# runs encoder and decoder
model(input_ids, decoder_input_ids)

# just run the encoder
encoder_outputs = model.encode(input_ids)

# run the decoder
decoder_outputs = model.decode(decoder_input_ids, encoder_outputs)
```

- For now `past_key_values` caching is only implemented in the self-attention layer in the decoder i.e it's not implemented for the cross-attention layer. 

<hr>

**Reviewers:** @patrickvonplaten @sgugger @patil-suraj (and whoever else in the community)

### CODE REVIEW

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 1]
```diff
@@ -341,6 +341,67 @@ def __call__(
         return outputs
 
 
+class FlaxBartEncoderLayerCollection(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:344]: This is needed to imitate the exact same weight structure as in PyTorch.

In PyTorch we always have things like `bart.encoder.layers.1` .... => this means that we need to create a class in Flax which represents `layers` and give each layer the weights name `1`, `2`, ...

File: tests/test_modeling_flax_common.py [hunk 2]
```diff
@@ -137,7 +137,7 @@ def recursive_check(tuple_object, dict_object):
     def test_equivalence_pt_to_flax(self):
         config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
 
-        for model_class in self.all_model_classes:
+        for model_class in self.all_model_classes[:-2]:  # TODO(Patrick, Daniel) - ForSeqClassification and QA doesn't work yet -> need to investigate
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_common.py:140]: For now I only made it work for `BartModel` & `BartForConditionalGeneration` -> should be easy to fix for `ForSeqClass....` and `ForQA....` as well, but didn't find time to investigate

**Comment by patrickvonplaten** [comment]
Hey @stancld,

One important thing we should implement before merging is the caching mechanismh similar to how it's done in GPT2: https://github.com/huggingface/transformers/blob/0b0a598452b02278075a75f84b5ca7bb457224ad/src/transformers/models/gpt2/modeling_flax_gpt2.py#L139. @patil-suraj, could you help here maybe? :-)

**Comment by patil-suraj** [comment]
Sure @patrickvonplaten !

@stancld  let me know if you need help here or I could take this if you are okay with it :)

File: src/transformers/generation_flax_utils.py [hunk 3]
```diff
@@ -101,12 +101,56 @@ def _run_loop_in_debug(cond_fn, body_fn, init_state):
...
+        return model_kwargs
+
+    def _prepare_decoder_input_ids_for_generation(
+        self, input_ids, decoder_start_token_id: int = None, bos_token_id: int = None
+    ):
+        decoder_start_token_id = self._get_decoder_start_token_id(decoder_start_token_id, bos_token_id)
+        decoder_input_ids = jnp.ones((input_ids.shape[0], 1), dtype="i4") * decoder_start_token_id
+        return decoder_input_ids
+
+    def _get_decoder_start_token_id(self, decoder_start_token_id: int = None, bos_token_id: int = None) -> int:
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:120]: actually to begin with can we make this super simple and just use `self.config.decoder_start_token_id`? If the model doesn't have a `decoder_start_token_id` then we'll raise an error. Don't want to introduce all those exceptions we have in PT here in the beginning as well

Comment (patrickvonplaten) [reply src/transformers/generation_flax_utils.py:120]: They are mostly for the EncoderDecoder class BTW

File: src/transformers/generation_flax_utils.py [hunk 4]
```diff
@@ -287,8 +345,10 @@ def greedy_search_body_fn(state):
                 model_kwargs=next_model_kwargs,
             )
 
-        # The very first prompt often has sequence length > 1, so run outside of `lax.while_loop` to comply with TPU
-        state = greedy_search_body_fn(state)
+        # for now, not accepting decoder_input_ids for seq2seq models, so we should only run the for causal models
+        if not self.config.is_encoder_decoder:
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:349]: Actually can be have the if statement, rather on whether `input_ids.shape > 1`? For Seq2Seq it could also be the case that `decoder_input_ids` to start generating from are provided

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 5]
```diff
@@ -0,0 +1,1394 @@
+# coding=utf-8
+# Copyright 2021 The Fairseq Authors, The HuggingFace Inc. team and Daniel Stancl. All rights reserved.
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:2]: think we sadly have to remove `Daniel Stancl` here :-/ We can't really have individual contributors in the Copyright I think (cc @sgugger )

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 6]
```diff
@@ -0,0 +1,1394 @@
...
+                key_states, value_states, query_states, attention_mask
+            )
+
+        # Convert the boolean attention mask to an attention bias.
+        if attention_mask is not None:
+            # attention mask in the form of attention bias
+            attention_bias = lax.select(
+                attention_mask > 0,
+                jnp.full(attention_mask.shape, 0.0).astype(self.dtype),
+                jnp.full(attention_mask.shape, float("-inf")).astype(self.dtype),
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:333]: do we have to use `-inf`? We use -1e4 in GPT2

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 7]
```diff
@@ -0,0 +1,1394 @@
...
+        dtype: jnp.dtype = jnp.float32,
+        **kwargs
+    ):
+        module = self.module_class(config=config, dtype=dtype, **kwargs)
+        super().__init__(config, module, input_shape=input_shape, seed=seed, dtype=dtype)
+
+    def init_weights(self, rng: jax.random.PRNGKey, input_shape: Tuple) -> FrozenDict:
+        # init input tensors
+        input_ids = jnp.zeros(input_shape, dtype="i4")
+        input_ids = jax.ops.index_update(
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:665]: why exactly do we need a EOS token for the init pass?

File: src/transformers/models/bart/modeling_flax_bart.py (hunk 5 above)
Comment (stancld) [reply src/transformers/models/bart/modeling_flax_bart.py]: That's okay, removed that! :)

File: src/transformers/models/bart/modeling_flax_bart.py (hunk 7 above)
Comment (stancld) [reply src/transformers/models/bart/modeling_flax_bart.py]: @patrickvonplaten At this moment, when one initializes a sequence classification model without an EOS token, there's an error raised by
```
if any(eos_mask.sum(1) == 0):
    raise ValueError("There are missing <eos> tokens in input_ids")
```
(This condition is drawn from BART's PyTorch implementation)

<hr>

I'm gonna slightly change a comment for the superior/preceding condition to this one above:
```
if type(eos_mask) != jax.interpreters.partial_eval.DynamicJaxprTracer:
```
which is actually there to ensure JIT compilation pass.

File: src/transformers/generation_flax_utils.py [hunk 8]
```diff
@@ -246,10 +272,9 @@ def _greedy_search(
         # per batch-item state bit indicating if sentence has finished.
         is_sent_finished = jnp.zeros((batch_size,), dtype=jnp.bool_)
 
-        model = self
-
+        model = self.decode if self.config.is_encoder_decoder else self
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:275]: this deserves a comment I think

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 9]
```diff
@@ -0,0 +1,1677 @@
...
+        all_attentions = () if output_attentions else None
+        all_hidden_states = () if output_hidden_states else None
+
+        for encoder_layer in self.layers:
+            if output_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
+            # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
+            dropout_probability = random.uniform(0, 1)
+            if deterministic and (dropout_probability < self.layerdrop):  # skip the layer
+                outputs = (None, None)
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:484]:
```suggestion
                outputs = (None, None)
            else:
                ...
```
or else layerdrop wolud never be actually applied no?

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 10]
```diff
@@ -0,0 +1,1677 @@
...
+        all_self_attns = () if output_attentions else None
+        all_cross_attentions = () if (output_attentions and encoder_hidden_states is not None) else None
+
+        for decoder_layer in self.layers:
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+                # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
+            dropout_probability = random.uniform(0, 1)
+            if deterministic and (dropout_probability < self.layerdrop):
+                continue
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:630]: for consistency, can we also just to `layer_outputs = (None, None,None)` here?

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 11]
```diff
@@ -0,0 +1,1677 @@
...
+            position_ids,
+            decoder_position_ids,
+        )["params"]
+
+    def init_cache(self, batch_size, max_length, encoder_outputs):
+        r"""
+        Args:
+            batch_size (:obj:`int`):
+                batch_size used for fast auto-regressive decoding. Defines the batch size of the initialized cache.
+            max_length (:obj:`int`):
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:978]: docstring for encoder_outputs missing

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 12]
```diff
@@ -0,0 +1,1677 @@
...
+        self.model = FlaxBartModule(config=self.config, dtype=self.dtype)
+        self.lm_head = nn.Dense(
+            self.model.shared.num_embeddings,
+            use_bias=False,
+            dtype=self.dtype,
+            kernel_init=jax.nn.initializers.normal(self.config.init_std, self.dtype),
+        )
+        self.final_logits_bias = self.param("final_logits_bias", self.bias_init, (1, self.model.shared.num_embeddings))
+
+    def get_encoder(self):
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:1198]: I'd prefer to delete those functions (not make them part of the public API) since there are a bit confusing to me `get_encoder()` actually returns a funtion here, no?

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 13]
```diff
@@ -0,0 +1,1677 @@
...
+        return self.model.encoder
+
+    def get_decoder(self):
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:1201]: same, I'd delete the function here

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 14]
```diff
@@ -0,0 +1,1677 @@
...
+@add_start_docstrings(
+    "The BART Model with a language modeling head. Can be used for summarization.", BART_START_DOCSTRING
+)
+class FlaxBartForConditionalGeneration(FlaxBartPretrainedModel):
+    module_class = FlaxBartForConditionalGenerationModule
+    dtype: jnp.dtype = jnp.float32
+
+    @add_start_docstrings(BART_DECODE_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=FlaxSeq2SeqLMOutput, config_class=BartConfig)
+    def decode(
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:1268]: nice!

File: tests/test_modeling_flax_common.py [hunk 15]
```diff
@@ -167,7 +178,8 @@ def test_equivalence_pt_to_flax(self):
                 fx_outputs = fx_model(**prepared_inputs_dict).to_tuple()
                 self.assertEqual(len(fx_outputs), len(pt_outputs), "Output lengths differ between Flax and PyTorch")
                 for fx_output, pt_output in zip(fx_outputs, pt_outputs):
-                    self.assert_almost_equals(fx_output, pt_output.numpy(), 4e-2)
+                    if not isinstance(fx_output, tuple):  # TODO(Patrick, Daniel) - let's discard use_cache for now
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_common.py:181]: Think we don't really need this no? I would always disable `use_cache` in PyTorch here instead `pt_model.config.use_cache = False` and leave the test as it is. In Flax the cache should not be returned as a default (only if `init_cache` is called which it is not here I think)

File: tests/test_modeling_flax_bart.py [hunk 16]
```diff
@@ -0,0 +1,406 @@
...
+            past_key_values=outputs_cache.past_key_values,
+            decoder_position_ids=decoder_position_ids,
+        )
+
+        outputs = model.decode(decoder_input_ids, encoder_outputs)
+
+        diff = np.max(np.abs((outputs_cache_next[0][:, -1, :5] - outputs[0][:, -1, :5])))
+        self.parent.assertTrue(diff < 1e-3, msg=f"Max diff is {diff}")
+
+    def check_use_cache_forward_with_attn_mask(self, model_class_name, config, inputs_dict):
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_bart.py:179]: very nice test!

**Comment by patil-suraj** [comment]
@patrickvonplaten  a couple of last questions about the API

- Right now the `FlaxBartPretrainedModel.__call__` method also accepts the `encoder_outputs`, but think it would be cleaner to not do that as we already have `encode/decode` methods and let the user use `encode` or `decode` if they want to run just one part of the model. So we could make `decode` method available for every model (right now it's only available for the `ForConditionalGeneration` model) and it'll return the decoder outputs and for `*ForConditionalGeneration` models, it'll also return the `logits`.

- The `decode` method returns `FlaxSeq2SeqLM` output, which includes both the encoder and decoder outputs, but when calling  `decode` the user already has `encoder_outputs`, so maybe we should just return decoder outputs since the `decode` method only runs the decoder.

What do you think?

**Comment by patrickvonplaten** [comment]
IMO:

1) Agree, happy to remove `encoder_outptus` as an input argument from `call` & make `decode` available for all models
2) Yes, it doesn't make too much sense to include all the encoder relevant output here!

**Review by patrickvonplaten — APPROVED** [review]

File: src/transformers/models/bart/modeling_flax_bart.py (hunk 6 above)
Comment (LysandreJik) [reply src/transformers/models/bart/modeling_flax_bart.py:365]: I would keep the same value as PyTorch GPT-2 as well

File: tests/test_modeling_flax_common.py [hunk 17]
```diff
@@ -207,8 +227,9 @@ def test_equivalence_flax_to_pt(self):
 
                 fx_outputs = fx_model(**prepared_inputs_dict).to_tuple()
                 self.assertEqual(len(fx_outputs), len(pt_outputs), "Output lengths differ between Flax and PyTorch")
+
                 for fx_output, pt_output in zip(fx_outputs, pt_outputs):
-                    self.assert_almost_equals(fx_output, pt_output.numpy(), 4e-2)
+                    self.assert_almost_equals(fx_output, pt_output.numpy(), 1e-3)
```
Comment (LysandreJik) [inline tests/test_modeling_flax_common.py:232]: Nice!

**Review by LysandreJik — APPROVED** [review]
Great! No problem for me to have a slight change of API to make it more Flax-centric with the `encode` and `decode`.

File: tests/test_modeling_flax_common.py [hunk 18]
```diff
@@ -167,7 +181,7 @@ def test_equivalence_pt_to_flax(self):
                 fx_outputs = fx_model(**prepared_inputs_dict).to_tuple()
                 self.assertEqual(len(fx_outputs), len(pt_outputs), "Output lengths differ between Flax and PyTorch")
                 for fx_output, pt_output in zip(fx_outputs, pt_outputs):
-                    self.assert_almost_equals(fx_output, pt_output.numpy(), 4e-2)
+                    self.assert_almost_equals(fx_output, pt_output.numpy(), 1e-3)
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_common.py:184]: This makes tests on current master fail - why did we change this!?

Comment (patrickvonplaten) [reply tests/test_modeling_flax_common.py:184]: Pushing a fix...
